pallets/inflation-manager/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+
+			Pallet::<T>::deposit_event(Event::InflationConfigurationSet {
+				inflation_configuration: self.inflation_configuration.clone(),
+			});
+		}
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(now: T::BlockNumber) {
+			// flag to check weather or not we're at end of a year and must recalculate block
+			// rewards
+			let mut logic_switch: bool = false;
+
+			let current_year = CurrentYear::<T>::get();
+			let inflation_config = InflationConfiguration::<T>::get();
+			let mut inflation_parameters = YearlyInflationParameters::<T>::get();
+			let mut block_rewards = BlockRewards::<T>::get();
+
+			// if we have reached the stagnation year, kill the recalculation flag
+			// and set inflation parameters to stagnation values
+			if current_year == inflation_config.inflation_stagnation_year {
+				inflation_parameters = InflationParametersT {
+					effective_inflation_rate: inflation_config.inflation_stagnation_rate,
+					effective_disinflation_rate: Perbill::one(),
+				};
+
+				// kill recalculation flag
+				RecalculationAt::<T>::kill();
+
+				// put stagnation inflation parameters into storage
+				YearlyInflationParameters::<T>::put(inflation_parameters.clone());
+
+				logic_switch = true;
+			}
+
+			// check if we need to recalculate inflation parameters for a new year
+			// update inflation parameters if we havent reached the stagnation year
+			if let Some(target_block) = RecalculationAt::<T>::get() {
+				if now >= target_block && current_year < inflation_config.inflation_stagnation_year
+				{
+					// update inflation parameters
+					inflation_parameters =
+						Self::update_inflation_parameters(&inflation_parameters.clone());
+					YearlyInflationParameters::<T>::put(inflation_parameters.clone());
+
+					// set the flag to calculate inflation parameters after a year(in blocks)
+					let target_block = now + T::BlockNumber::from(BLOCKS_PER_YEAR);
+					RecalculationAt::<T>::put(target_block);
+
+					logic_switch = true;
+				}
+			}
+
+			// recalculate block rewards and update current year if inflation parameters were
+			// updated
+			if logic_switch {
+				// calculate new block rewards
+				block_rewards = Self::rewards_per_block(&inflation_parameters);
+
+				// put new block rewards into storage
+				BlockRewards::<T>::put(block_rewards);
+
+				// log this change
+				Self::deposit_event(Event::BlockRewardsUpdated { block_rewards });
+
+				// update current year in any case
+				CurrentYear::<T>::put(current_year + 1);
+			}
+
+			// Event
+			Self::deposit_event(Event::InflationParametersUpdated {
+				current_year,
+				inflation_parameters: inflation_parameters.clone(),
+				block_rewards,
+			});
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> Pallet<T> {
+		// calculate inflationary tokens per block
+		fn rewards_per_block(inflation_parameters: &InflationParametersT) -> Balance {
```

Comment:
In my understanding, we'll calculate the rewards_per_block once a year and store it in the blockchain. The block reward pallet will get the information from the blockchain for token minting, right?

Reply:
BlockRewards are updated at end of the year, and the value is stored as a StorageValue in inflation-manager pallet, block reward pallet then fetches the value from there for minting.